WordPress/openverse
# Always use only latest provider tags when reingesting

## Fixes



Fixes #4732 by @sarayourfriend 

## Description


Changes the tag upsert strategy to match the requirements described in the issue. Namely: provider tags must always reflect the latest tags from the provider, and never retain deleted or changed tags in our data.

The PR does this by adding a new column upsert strategy exclusively for the tags column which filters the provider tags out from the column's existing tags before merging it with the incoming tags. This has the effect of preserving all non-provider tags whilst simply ignoring any previously held provider tags in favour of always only using the incoming tags.

## Testing Instructions

CI and unit tests should pass, of course. Review the changes I made to the test and ensure they are clear at explaining the intention of how the tags merging should work, and that I've covered the edge cases.

You can also manually test. Run Airflow with `ov just catalog/up` and visit localhost:9090 and log in with airflow as the username and password.

Add the `INGESTION_LIMIT` variable through the Admin variable page (http://0.0.0.0:9090/variable/list/) and set it to a low number like 10.

Enable and let the `wordpress_workflow` DAG run (http://0.0.0.0:9090/dags/wordpress_workflow/grid). You will now have 10 WordPress Photo Directory results in your local data.

Run `ov just catalog/pgcli` and execute the following SQL to mock some enriched tags as well as tags that "used to be" in the provider (and which will disappear after reingestion):

```sql
update image set tags = tags || '[{"name": "deleted_upstream", "provider": "wordpress"}, {"name": "chicken", "provider": "magic_vision"}]'::jsonb where identifier in (select identifier from image where provider = 'wordpress' limit 1) returning identifier, tags;
```

Record the identifier returned by the query.

Then re-run the WordPress workflow DAG by manually running it (http://0.0.0.0:9090/dags/wordpress_workflow/trigger?origin=%2Fdags%2Fwordpress_workflow%2Fgrid). You do not need to change the configuration at all. Let it finish running, then in the pgcli, select the tags for the image you modified before (use the identifier returned by the `update` query). You should see the "chicken" tag retained (because it is from "magic_vision" rather than "wordpress" provider), but should no longer see "deleted_upstream" in the tags list, because the incoming tags when you reingested no longer included those. Confirm this is the case.

## Checklist


- [N/A] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
